exporter/elasticsearchexporter/config.go

```diff
@@ -0,0 +1,217 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package elasticsearchexporter
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"go.opentelemetry.io/collector/config/configmodels"
+	"go.opentelemetry.io/collector/config/configtls"
+)
+
+// Config defines configuration for Elastic exporter.
+type Config struct {
+	configmodels.ExporterSettings `mapstructure:",squash"`
+
+	// URLs holds the Elasticsearch URLs the exporter should send events to.
+	//
+	// This setting is required if CloudID is not set and if the
+	// ELASTICSEARCH_URL environment variable is not set.
+	URLs []string `mapstructure:"urls"`
+
+	// CloudID holds the cloud ID to identify the Elastic Cloud cluster to send events to.
+	// https://www.elastic.co/guide/en/cloud/current/ec-cloud-id.html
+	//
+	// This setting is required if no URL is configured.
+	CloudID string `mapstructure:"cloudid"`
+
+	// Workers configures the number of workers publishing bulk requests.
+	Workers int `mapstructure:"workers"`
+
+	// Index configures the index, index alias, or data stream name events should be indexed in.
+	//
+	// https://www.elastic.co/guide/en/elasticsearch/reference/current/indices.html
+	// https://www.elastic.co/guide/en/elasticsearch/reference/current/data-streams.html
+	//
+	// This setting is required.
+	Index string `mapstructure:"index"`
+
+	// Pipeline configures the ingest node pipeline name that should be used to process the
+	// events.
+	//
+	// https://www.elastic.co/guide/en/elasticsearch/reference/current/ingest.html
+	Pipeline string `mapstructure:"pipeline"`
+
+	HTTPClientSettings `mapstructure:",squash"`
```

Comment:
Consider using https://github.com/open-telemetry/opentelemetry-collector/tree/main/config/confighttp, so that this exporter is configured just like other HTTP exporters. One thing to pay attention to is that it includes a property for the `endpoint`, so, you'd have to remove the URLs property from this here.

Reply:
This structure is modeled after `confighttp.HTTPClientSettings`. I didn't use it though as the Elasticsearch exporter connects to a cluster and not a single endpoint (`urls` setting). For self-hosted users it is quite common to configure the multiple URLs for load-balancing and failover in case some node is not available. Even with discovery enabled one might want to configure multiple seed endpoints for discoverability, in which case the setting is more similar to the `brokers` setting in the Kafka exporter.

It seems `endpoint` is more common thanks to the `HTTPClientSettings`. I think I will rename the `urls` setting to `endpoints` in order to have a much more consistent configuration.

But I would prefer to not use `confighttp.HTTPClientSettings`, because it introduce a setting we do not support. If possible I would like `confighttp.HTTPClientSettings` to remove `endpoint`, or have a special `confighttp.HTTPTransportSettings` struct without `endpoint`. Even the `(*HTTPClientSettings).ToClient()` method ignores endpoint, as for the `net/http` API the `endpoint` is part of the request, not a constant setting in the Client.

Will open a PR to introduce `confighttp.HTTPTransportSettings`, let's see how it goes. In general +1 on reusing a common definition here.

Reply:
Sounds good to me!

Reply:
PR introducing `HTTPTransportSettings`: https://github.com/open-telemetry/opentelemetry-collector/pull/2470